File: sorting/quicksort.c
#include <stdio.h>

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

int partition(int array[], int left_idx, int right_idx) {
  
  int pivot = array[right_idx];
  int i = (left_idx - 1);

  for (int j = left_idx; j < right_idx; j++) {
    if (array[j] <= pivot) {
      i++;
      swap(&array[i], &array[j]);
    }
  }

  swap(&array[i + 1], &array[right_idx]);
  return (i + 1);
}

void quick_sort(int array[], int left_idx, int right_idx) {
  if (left_idx < right_idx) {
    int p = partition(array, left_idx, right_idx);
    quick_sort(array, left_idx, p - 1);
    quick_sort(array, p + 1, right_idx);
  }
}

File: sorting/quicksort.py
'''Quick Sort algorithm '''


def partition(array, left_idx,right_idx):
    pivot = array[right_idx]
    i = left_idx - 1

    for j in range(left_idx,right_idx):
        if array[j] <= pivot:
            i += 1
            array[i],array[j] = array[j],array[i]

    array[i + 1],array[right_idx] = array[right_idx],array[i + 1]
    return i + 1


def quick_sort(array,left_idx,right_idx):
    if left_idx < right_idx:
        p = partition(array,left_idx,right_idx)
        quick_sort(array,left_idx, p - 1)
        quick_sort(array,p + 1,right_idx)
